File: packages/react-native-storybook/ios/LastStateHelper.m
#import "LastStateHelper.h"
#import "FileSystemHelper.h"

static NSString *const LOG_TAG = @"SherloModule:LastStateHelper";
static NSString *const PROTOCOL_FILENAME = @"protocol.sherlo";

/**
 * Helper for managing the last state of the Sherlo module.
 * Provides functionality to save, load, and retrieve the last state 
 * of the application, such as the active mode and other session-specific data.
 */
@implementation LastStateHelper

/**
 * Reads the protocol file and extracts the last state information.
 * This includes the next snapshot index, filtered view IDs, and request ID.
 * 
 * @param fileSystemHelper The file system helper for reading files
 * @return NSDictionary containing the state information or nil if not found
 */
+ (NSDictionary *)getLastState:(FileSystemHelper *)fileSystemHelper {
  @try {
    NSError *readError = nil;
    NSString *protocolContent = [fileSystemHelper readFile:PROTOCOL_FILENAME error:&readError];

    if (readError || !protocolContent || [protocolContent stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]].length == 0) {
      NSLog(@"[%@] Protocol file is empty or doesn't exist", LOG_TAG);
      return [NSDictionary dictionary];
    }

    // Try to decode base64 if content is base64 encoded
    @try {
      NSData *decodedData = [[NSData alloc] initWithBase64EncodedString:protocolContent options:0];
      if (decodedData) {
        NSString *decodedString = [[NSString alloc] initWithData:decodedData encoding:NSUTF8StringEncoding];
        if (decodedString) {
          protocolContent = decodedString;
        }
      }
    } @catch (NSException *exception) {
      NSLog(@"[%@] Content is not base64 encoded, using raw content", LOG_TAG);
    }

    NSArray *responseLines = [protocolContent componentsSeparatedByString:@"\n"];
    NSDictionary *ackStart = nil;
    NSDictionary *lastRequestSnapshot = nil;
    NSDictionary *startItem = nil;  // Variable to store START action

    // Iterate through all lines in reverse order
    for (NSInteger i = responseLines.count - 1; i >= 0; i--) {
      @try {
        NSString *line = responseLines[i];
        if (!line || [line stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]].length == 0) {
          continue;
        }

        NSData *jsonData = [line dataUsingEncoding:NSUTF8StringEncoding];
        NSDictionary *responseItem = [NSJSONSerialization JSONObjectWithData:jsonData options:0 error:&readError];

        if (readError) {
          NSLog(@"[%@] Error parsing protocol line: %@", LOG_TAG, readError.localizedDescription);
          continue;
        }

        NSString *action = responseItem[@"action"];
        if (!action) {
          continue;
        }

        if ([action isEqualToString:@"ACK_START"] && !ackStart) {
          ackStart = responseItem;
        } else if ([action isEqualToString:@"ACK_REQUEST_SNAPSHOT"] && !lastRequestSnapshot) {
          lastRequestSnapshot = responseItem;
        } else if ([action isEqualToString:@"START"] && !startItem) {
          startItem = responseItem;
        }

        // If we found all items, we can stop searching
        if (ackStart && lastRequestSnapshot && startItem) {
          break;
        }
      } @catch (NSException *exception) {
        // Ignore parse errors for invalid JSON lines
        NSLog(@"[%@] Error parsing protocol line: %@", LOG_TAG, exception.reason);
        continue;
      }
    }

    NSMutableDictionary *state = [NSMutableDictionary dictionary];
    if (ackStart) {
      NSDictionary *nextSnapshot;
      if (lastRequestSnapshot && lastRequestSnapshot[@"nextSnapshot"]) {
        nextSnapshot = lastRequestSnapshot[@"nextSnapshot"];
      } else if (ackStart[@"nextSnapshot"]) {
        nextSnapshot = ackStart[@"nextSnapshot"];
      } else {
        nextSnapshot = @{};
      }
      
      [state setObject:nextSnapshot forKey:@"nextSnapshot"];

      NSString *requestId = @"";
      if (lastRequestSnapshot && lastRequestSnapshot[@"requestId"]) {
        requestId = lastRequestSnapshot[@"requestId"];
      } else if (ackStart[@"requestId"]) {
        requestId = ackStart[@"requestId"];
      }

      [state setObject:requestId forKey:@"requestId"];
    }

    return state;
  } @catch (NSException *exception) {
    NSLog(@"[%@] Error getting last state: %@", LOG_TAG, exception.reason);
    return [NSDictionary dictionary];
  }
}

@end 

File: packages/react-native-storybook/android/src/main/java/io/sherlo/storybookreactnative/LastStateHelper.java
package io.sherlo.storybookreactnative;

import android.util.Log;
import org.json.JSONObject;
import org.json.JSONArray;
import android.util.Base64;

/**
 * Helper for extracting and processing state information from Sherlo protocol files.
 * Used to maintain continuity between test runs or application restarts.
 */
public class LastStateHelper {
    private static final String TAG = "LastStateHelper";
    private static final String PROTOCOL_FILENAME = "protocol.sherlo";
    
    /**
     * Reads the protocol file and extracts the last state information.
     * Processes the file to find the most recent relevant protocol entries and
     * constructs a state object with snapshot and request information.
     * 
     * @param fileSystemHelper The file system helper for reading the protocol file
     * @return JSONObject containing the extracted state information or an empty JSONObject if not found
     */
    public static JSONObject getLastState(FileSystemHelper fileSystemHelper) {
        try {
            String protocolContent = null;
            try {
                protocolContent = fileSystemHelper.readFile(PROTOCOL_FILENAME);
            } catch (Exception e) {
                Log.w(TAG, "Protocol file doesn't exist or is not readable: " + e.getMessage());
                return new JSONObject();
            }
            
            if (protocolContent == null || protocolContent.trim().isEmpty()) {
                Log.w(TAG, "Protocol file is empty");
                return new JSONObject();
            }

            // Try to decode if base64 encoded
            try {
                byte[] decodedBytes = Base64.decode(protocolContent, Base64.DEFAULT);
                protocolContent = new String(decodedBytes, "UTF-8");
            } catch (Exception e) {
                Log.w(TAG, "Protocol is not base64 encoded, trying to parse as plain JSON");
            }

            String[] responseLines = protocolContent.split("\n");
            JSONObject ackStart = null;
            JSONObject lastRequestSnapshot = null;
            JSONObject startItem = null;  // Variable to store START action

            // Iterate through all lines in reverse order
            for (int i = responseLines.length - 1; i >= 0; i--) {
                try {
                    String line = responseLines[i];
                    if (line == null || line.trim().isEmpty()) continue;

                    JSONObject responseItem = new JSONObject(line);
                    String action = responseItem.optString("action", "");

                    if ("ACK_START".equals(action) && ackStart == null) {
                        ackStart = responseItem;
                    } else if ("ACK_REQUEST_SNAPSHOT".equals(action) && lastRequestSnapshot == null) {
                        lastRequestSnapshot = responseItem;
                    } else if ("START".equals(action) && startItem == null) {
                        startItem = responseItem;
                    }

                    // If we found all items, we can stop searching
                    if (ackStart != null && lastRequestSnapshot != null && startItem != null) {
                        break;
                    }
                } catch (Exception e) {
                    // Ignore parse errors for invalid JSON lines
                    Log.w(TAG, "Error parsing protocol line: " + e.getMessage());
                    continue;
                }
            }

            JSONObject state = new JSONObject();
            if (ackStart != null) {
                JSONObject nextSnapshot;
                if (lastRequestSnapshot != null && lastRequestSnapshot.has("nextSnapshot")) {
                    nextSnapshot = lastRequestSnapshot.getJSONObject("nextSnapshot");
                } else if (ackStart.has("nextSnapshot")) {
                    nextSnapshot = ackStart.getJSONObject("nextSnapshot");
                } else {
                    nextSnapshot = new JSONObject();
                }

                state.put("nextSnapshot", nextSnapshot);

                String requestId = "";
                if (lastRequestSnapshot != null && lastRequestSnapshot.has("requestId")) {
                    requestId = lastRequestSnapshot.getString("requestId");
                } else if (ackStart.has("requestId")) {
                    requestId = ackStart.getString("requestId");
                }

                state.put("requestId", requestId);
            }

            return state;
        } catch (Exception e) {
            Log.e(TAG, "Error getting last state", e);
            return new JSONObject();
        }
    }
} 